### PR TITLE
[IMP] {website_}event: propagate event description to calendar 

### PR DESCRIPTION
**Pull request by thsh-odoo**
After this commit if we add the event using the add to
calendar buttons we also propagate the link of the event
and the description of the event.

Note -: The link of the event will only propagate in the description
if the website_event module is installed.

Technical
=========
When we add the event to a calendar using Gcal button
we propagate the event data through link so to prevent the
crash if the description is too long we added the max url
length.

Reference docs for the URL limit:
https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers

Task-3861438

### CODE REVIEW

File: addons/event/models/event_event.py [hunk 1]
```diff
@@ -10,7 +10,7 @@
 from odoo.addons.base.models.res_partner import _tz_get
 from odoo.exceptions import UserError, ValidationError
 from odoo.osv import expression
-from odoo.tools import format_datetime, is_html_empty
+from odoo.tools import format_datetime, is_html_empty, html2plaintext
```
Comment (ibo-odoo) [inline addons/event/models/event_event.py:13]: @thsh-odoo Simply rearrange it in alphabetical order, and everything else looks good to proceed.

File: addons/event/models/event_event.py [hunk 2]
```diff
@@ -644,6 +646,13 @@ def mail_attendees(self, template_id, force_send=False, filter_func=lambda self:
             for attendee in event.registration_ids.filtered(filter_func):
                 self.env['mail.template'].browse(template_id).send_mail(attendee.id, force_send=force_send)
 
+    def _get_event_description(self):
```
Comment (awa-odoo) [inline addons/event/models/event_event.py:649]: Sugg: "_get_external_description"

File: addons/event/models/event_event.py [hunk 3]
```diff
@@ -644,6 +646,13 @@ def mail_attendees(self, template_id, force_send=False, filter_func=lambda self:
...
+        """ Returns the event description with a link to the event registration page. """
```
Comment (awa-odoo) [inline addons/event/models/event_event.py:650]: Sugg:

```py
""" Description & URL of the event shortened to maximum 2000 characters.
Meant to be used for external content (ics/icalc). """
```

Would also be nice to justify the 2000 characters limit somehow, by mentioning external provider or browsers documentations of such limits for example.

File: addons/event/models/event_event.py [hunk 4]
```diff
@@ -644,6 +646,13 @@ def mail_attendees(self, template_id, force_send=False, filter_func=lambda self:
...
+        max_length = 2000
+        event_url = Markup('<a href="%s/event/%s/register">%s</a>') % (self.get_base_url(), slug(self), self.name)
```
Comment (awa-odoo) [inline addons/event/models/event_event.py:652]: I don't really understand why you need Markup here.

More importantly, this URL only becomes available when "website_event" is installed.
So either we remove the URL, either this methods needs to be overridden in website_event somehow.

**Review by awa-odoo — COMMENTED** [review]
Hello @thsh-odoo 

A few comments on this one.

Thanks & cheers!

File: addons/event/models/event_event.py (hunk 4 above)
Comment (thsh-odoo) [reply addons/event/models/event_event.py:652]: Hello @awa-odoo 
I used markup because I thought we always need to markup the HTML content :sweat_smile: 
https://www.odoo.com/documentation/17.0/developer/reference/backend/security.html#creating-safe-content-using-markup
I will look around point 2.

File: addons/event/models/event_event.py (hunk 3 above)
Comment (thsh-odoo) [reply addons/event/models/event_event.py]: Hello @awa-odoo 
I wasn't able to find the official documentation of this URL limit so I have put the documentation from which I have taken the reference

**Comment by thsh-odoo** [comment]
Hello @awa-odoo 
Changes are done can you please review it again 
and I have a question functionally I have it to you and LNA on the Log note
Thanks :smile:

File: addons/website_event/models/event_event.py [hunk 5]
```diff
@@ -507,6 +507,13 @@ def _track_subtype(self, init_values):
             return self.env.ref('website_event.mt_event_unpublished', raise_if_not_found=False)
         return super(Event, self)._track_subtype(init_values)
 
+    def _get_external_description(self):
+        # Adding the URL of the event into the description
+        max_length = 2000
+        event_url = f'<a href="{self.get_base_url()}/event/{slug(self)}/register">{self.name}</a>'
```
Comment (awa-odoo) [inline addons/website_event/models/event_event.py:513]: Could use "self.website_url" here?

Would add self.ensure_one as well.

File: addons/website_event/models/event_event.py [hunk 6]
```diff
@@ -507,6 +507,13 @@ def _track_subtype(self, init_values):
...
+        description = event_url + '\n' + super()._get_external_description()
+        return (description[:max_length] + '...') if len(description) > max_length else description
```
Comment (awa-odoo) [inline addons/website_event/models/event_event.py:515]: Seems a bit overkill to do that whole thing twice.

Would say we either reduce the limit in the other method to 1900 to leave space for that extra URL or live with the fact that it will be 2000 + url size.

Depends a bit on the hard limit.

File: addons/event/models/event_event.py [hunk 7]
```diff
@@ -644,6 +644,16 @@ def mail_attendees(self, template_id, force_send=False, filter_func=lambda self:
...
                 self.env['mail.template'].browse(template_id).send_mail(attendee.id, force_send=force_send)
 
+    def _get_external_description(self):
+        """ Description of the event shortened to maximum 2000 characters.
+        Meant to be used for external content (ics/icalc/Gcal).
+
+        Reference Docs for URl limit -: https://wiert.me/2010/04/20/maximum-url-lengths/
+        """
+        max_length = 2000
+        description = html_to_inner_content(self.description)
```
Comment (awa-odoo) [inline addons/event/models/event_event.py:654]: Needs "self.ensure_one".

File: addons/event/models/event_event.py [hunk 8]
```diff
@@ -644,6 +644,16 @@ def mail_attendees(self, template_id, force_send=False, filter_func=lambda self:
             for attendee in event.registration_ids.filtered(filter_func):
                 self.env['mail.template'].browse(template_id).send_mail(attendee.id, force_send=force_send)
 
+    def _get_external_description(self):
+        """ Description of the event shortened to maximum 2000 characters.
+        Meant to be used for external content (ics/icalc/Gcal).
+
+        Reference Docs for URl limit -: https://wiert.me/2010/04/20/maximum-url-lengths/
```
Comment (awa-odoo) [inline addons/event/models/event_event.py:651]: Seems a bit outdated, but if we don't have better... I'll take it :smile:

**Review by awa-odoo — COMMENTED** [review]
Hello @thsh-odoo 

Some more comments on this one.

For the functional part, I let Luc answer :smile: 

Thanks & cheers!

File: addons/event/models/event_event.py (hunk 8 above)
Comment (thsh-odoo) [reply addons/event/models/event_event.py:651]: Hello @awa-odoo 
it is not documentation but I think it is answered by the Google AdWords API Team 
https://groups.google.com/g/adwords-api/c/1pUlvZQK58I/m/VPKLiZldAgAJ?pli=1

can we put this?  :smile:

**Comment by thsh-odoo** [comment]
Hello @awa-odoo 
Changes are done.
Can you please review it again :smiley: 

Thanks!!

File: addons/event/models/event_event.py (hunk 8 above)
Comment (awa-odoo) [reply addons/event/models/event_event.py:651]: Let's put this one actually:
https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers

Even more direct.

(To adapt in the commit message as well).

File: addons/event/models/event_event.py [hunk 9]
```diff
@@ -644,6 +644,17 @@ def mail_attendees(self, template_id, force_send=False, filter_func=lambda self:
...
+        """
+        self.ensure_one()
+        max_length = 2000
+        description = html_to_inner_content(self.description)
+        return (description[:max_length] + '...') if len(description) > max_length else description
```
Comment (awa-odoo) [inline addons/event/models/event_event.py:656]: Could we use "textwrap.shorten" for this or does it mess-up the new lines?

File: addons/website_event/models/event_event.py [hunk 10]
```diff
@@ -507,6 +507,14 @@ def _track_subtype(self, init_values):
             return self.env.ref('website_event.mt_event_unpublished', raise_if_not_found=False)
         return super(Event, self)._track_subtype(init_values)
 
+    def _get_external_description(self):
+        # Adding the URL of the event into the description
```
Comment (awa-odoo) [inline addons/website_event/models/event_event.py:511]: Might as well put this as a docstring then.

File: addons/website_event/models/event_event.py [hunk 11]
```diff
@@ -507,6 +507,14 @@ def _track_subtype(self, init_values):
...
+        self.ensure_one()
+        max_length = 1900
```
Comment (awa-odoo) [inline addons/website_event/models/event_event.py:513]: As said before, would say either put the initial max_length at 1900 or keep 2000 + the URL here.

If each module starts re-defining a new max_length, it becomes a bit hard to follow.
Will also simplify the code, as this just needs to append the URL and be done.

File: addons/website_event/models/event_event.py [hunk 12]
```diff
@@ -507,6 +507,14 @@ def _track_subtype(self, init_values):
...
+        event_url = f'<a href="{self.get_base_url()}{self.website_url}/register">{self.name}</a>'
```
Comment (awa-odoo) [inline addons/website_event/models/event_event.py:514]: A bit better to use "url_join" to combine the various path.

**Review by awa-odoo — COMMENTED** [review]
Hello @thsh-odoo 

A few more comments on this one.

Thanks & cheers!

File: addons/website_event/models/event_event.py (hunk 12 above)
Comment (thsh-odoo) [reply addons/website_event/models/event_event.py:514]: Hello @awa-odoo 
rather than joining the various path I use this field `event_register_url` which contains the proper URL that I am making if that's okay to you

File: addons/event/models/event_event.py [hunk 13]
```diff
@@ -644,6 +645,17 @@ def mail_attendees(self, template_id, force_send=False, filter_func=lambda self:
...
+    def _get_external_description(self):
+        """ Description of the event shortened to maximum 2000 characters.
+        Meant to be used for external content (ics/icalc/Gcal).
+
+        Reference Docs for URl limit -: https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers
+        """
+        self.ensure_one()
+        max_length = 1900
+        description = html_to_inner_content(self.description)
+        return textwrap.shorten(description + '...', max_length)
```
Comment (awa-odoo) [inline addons/event/models/event_event.py:657]: "textwrap" already adds the "..." if necessary, no need to add it manually.

File: addons/event/models/event_event.py [hunk 14]
```diff
@@ -644,6 +645,17 @@ def mail_attendees(self, template_id, force_send=False, filter_func=lambda self:
...
                 self.env['mail.template'].browse(template_id).send_mail(attendee.id, force_send=force_send)
 
+    def _get_external_description(self):
+        """ Description of the event shortened to maximum 2000 characters.
+        Meant to be used for external content (ics/icalc/Gcal).
+
+        Reference Docs for URl limit -: https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers
+        """
+        self.ensure_one()
+        max_length = 1900
```
Comment (awa-odoo) [inline addons/event/models/event_event.py:655]: Would not make it a variable, not necessary.

File: addons/event/models/event_event.py [hunk 15]
```diff
@@ -644,6 +645,17 @@ def mail_attendees(self, template_id, force_send=False, filter_func=lambda self:
             for attendee in event.registration_ids.filtered(filter_func):
                 self.env['mail.template'].browse(template_id).send_mail(attendee.id, force_send=force_send)
 
+    def _get_external_description(self):
+        """ Description of the event shortened to maximum 2000 characters.
+        Meant to be used for external content (ics/icalc/Gcal).
+
```
Comment (awa-odoo) [inline addons/event/models/event_event.py:651]: Would mention that we actually shorten it to 1900 to leave some space for addition by sub-modules, such as the even link.

File: addons/event/models/event_event.py [hunk 16]
```diff
@@ -644,6 +645,17 @@ def mail_attendees(self, template_id, force_send=False, filter_func=lambda self:
...
+        Reference Docs for URl limit -: https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers
```
Comment (awa-odoo) [inline addons/event/models/event_event.py:652]: "URl" -> "URL"

**Review by awa-odoo — COMMENTED** [review]
@thsh-odoo 

Some last polishing details 🙂 

Thanks & cheers!

**Comment by thsh-odoo** [comment]
Hello @awa-odoo 
Changes are done.
Thanks!! :smile:

**Comment by awa-odoo** [comment]
@robodoo r+

**Comment by awa-odoo** [comment]
@robodoo r+
